dwave/optimization/include/dwave-optimization/nodes/constants.hpp

```diff
@@ -118,16 +129,54 @@ class ConstantNode : public ArrayOutputMixin<ArrayNode> {
     // holds its values on the object itself rather than in a State.
     double* buffer_ptr_;
 
-    // Information about the values in the buffer
-    struct BufferStats {
-        BufferStats() = delete;
-        explicit BufferStats(std::span<const double> buffer);
-
-        bool integral;
-        double min;
-        double max;
-    };
     mutable std::optional<BufferStats> buffer_stats_;
 };
 
+// InputNode acts like a placeholder or store of data very similar to ConstantNode,
+// with the key different being that its contents *may* change in between propagations.
+// However, it is not a decision variable--instead its use cases are acting as an "input"
+// for "models as functions", or for placeholders in large models where (otherwise constant)
+// data changes infrequently (e.g. a scheduling problem with a preference matrix).
+//
+// Initial values can be provided at node construction that will be used on state initialization.
+// Otherwise, the node will be initialized with all zeros.
+// TODO: finalize the constructors
+class InputNode : public ConstantNode {
```

Comment:
It is not clear to me what benefit this gets from subclassing `ConstantNode`? It holds a contiguous buffer it seems so it could be a lot simpler - constant does a lot of other things. I was imagining something like
```c++
class InputNode : ArrayOutputMixin<ArrayNode> {

    void initialize_state(State& state) {
        throw std::logic_error("input nodes must have their state initialized explicitly");
    }
    void initialize_state(State& state, std::span<double> values);
}
```
This means that one must do something like
```c++
auto state = graph.empty(state);
input_ptr->initialize_state(state, ...);
graph.initialize_state(state);  // finish everything else
```
obviously for the "outer" graph we would want to check before upload that there are no `Input` symbols, or maybe we just initialize them to `0` by default.

Reply:
My thinking was that

1. We need the `InputNode` to have a default value so that you _don't_ have to initialize it explicitly.
2. Given 1, might as well take advantage of the constructors of `ConstantNode`

If we can relax 1., then I agree, this would be much simpler! I think my original reasoning for that requirement was flawed anyway.